### PR TITLE
All three brackets for variables

### LINKED ISSUE #14464: [Theia AI] Allows three brackets for variables in prompt templates

### PR DESCRIPTION
**Pull request by JonasHelming**
fixed #14464

#### What it does

Allows to use three brackets for variables in prompts. No mixed formats are allows for consistency

#### How to test

There are automated test.
To manually test, 
- open the code completion prompt and a code file. 
- Add one bracket to {{textUntilCurrentPosition}}, see how it fails. 
- Add the second bracket to complete {{{textUntilCurrentPosition}}} and see how it works again. 
- Check that the variable is still listed in the AI configuration view then.
- Check that in the prompt editor, three brackets are valid, but mixtures are an error

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/ai-core/data/prompttemplate.tmLanguage.json [hunk 1]
```diff
@@ -49,4 +88,4 @@
   "fileTypes": [
     ".prompttemplate"
   ]
-}
+}
```
Comment (planger) [inline packages/ai-core/data/prompttemplate.tmLanguage.json:91]:
```suggestion
}

```

File: packages/ai-core/src/common/prompt-service.ts [hunk 2]
```diff
@@ -181,13 +181,18 @@ export class PromptServiceImpl implements PromptService {
...
+    }
+
     async getPrompt(id: string, args?: { [key: string]: unknown }): Promise<ResolvedPromptTemplate | undefined> {
         const prompt = this.getRawPrompt(id);
         if (prompt === undefined) {
             return undefined;
         }
 
-        const matches = [...prompt.template.matchAll(PROMPT_VARIABLE_REGEX)];
+        const matches = this.matchVariables(prompt.template);
```
Comment (planger) [inline packages/ai-core/src/common/prompt-service.ts:195]: Is there a reason why we extract this call of `matchVariablesRegEx()` into a method `this.matchVariables()`? Is this to simplify overriding it by adopters? Not sure this is a very common use case though.

I personally would keep this function call direct and not extract it into a separate method in this class (similar to how we do it in `agent-configuration-widget.tsx`.

If you like to keep the method (`this.matchVariables()`), I wouldn't make it `public` but rather `protected`; if we really want to keep the method at all.

Comment (planger) [reply packages/ai-core/src/common/prompt-service.ts:195]: Should we also put the other invocation of `prompt.template.matchAll` into a function for consistency?
See `[...prompt.template.matchAll(PROMPT_FUNCTION_REGEX)]` on line 215.

**Review by planger — CHANGES_REQUESTED** [review]
Thanks, interpreting the variables seem to work fine!

However, the syntax highlighting doesn't work for me:
image

Also I have a few comments inline above.

File: packages/ai-core/src/common/prompt-service.spec.ts
Comment (planger) [inline packages/ai-core/src/common/prompt-service.spec.ts:1]: Sorry, this message somehow got lost, readding it:

Nice, thanks for adding additional tests also covering the negative cases now.

I think we may want to avoid duplicating the three tests for the two- and three-bracket cases. This will reduce maintenance effort in the future. I think the following should work (but haven't tested it):

```ts
['{{', '{{{'].forEach((openingBrackets, index) => {
    const closingBrackets = openingBrackets.replace(/{/g, '}'); // Replace each `{` with `}`
    it(`should retrieve raw prompt by id (${index + 2} brackets)`, () => {
        const rawPrompt = promptService.getRawPrompt('8');
        expect(rawPrompt?.template).to.equal(`Hello, ${openingBrackets}name${closingBrackets}`);
    });
});
```

**Comment by JonasHelming** [comment]
> However, the syntax highlighting doesn't work for me

Should be fixed

File: packages/ai-core/src/common/prompt-service.spec.ts
Comment (JonasHelming) [reply packages/ai-core/src/common/prompt-service.spec.ts:1]: I am no sure about this proposal. I agree that the duplication on the tests might require maintenance (if we ever change this). However, I am not a huge fan of making tests with loops instead of clearly writing them out. I think if the test ever break, it will be much easier to understand the test cases if they are simple.

**Review by planger — APPROVED** [review]
Thanks for the revision, works fine now! :+1:
